lib/kdl/include/kdl/overloaded.h

```diff
@@ -0,0 +1,39 @@
+/*
+ Copyright 2010-2019 Kristian Duske
+
+ Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+ documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+ rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ persons to whom the Software is furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+ Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+ WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+#ifndef TRENCHBROOM_OVERLOADED_H
+#define TRENCHBROOM_OVERLOADED_H
+
+namespace kdl {
+    /**
+     * Creates a type that inherits from all of its type parameters and `operator()` from each supertype. Can be used
+     * with std::visit to create on-the-fly visitors from lambdas. So usually, the supertypes are lambdas.
```

Comment:
It was used for a bit in `AABBTree::LambdaVisitor` to get rid of `std::function`. I'm now using a template there because then the GCC bug about nested deduction guides is easier to work around. This is a very useful way to build visitors on the fly, but our code doesn't lend itself very well to this (yet). Actually I'd like to try to change our visitor pattern usages in a way that plays well with this, because we could use it to get rid if some of the node visitor classes.

Reply:
Cool. Yeah, that sounds good- the Visitor pattern can feel very heavyweight and while it's sometimes needed, often a simple lambda can replace it